### ISSUE TITLE
ResolutionImpossible for `numpy==1.21.5 ` for `requires-python = "*"`

### ISSUE THREAD

**Issue by gregdan3** [pdm-project/pdm#818]
- [X] I have searched the issue tracker and believe that this is not a duplicate.

**Make sure you run commands with `-v` flag before pasting the output.**

## Steps to reproduce

1. `pdm init` and follow the prompts
2. `pdm add numpy==1.21.5`

## Actual behavior
numpy fails to install with the following error:
```
~/temp2 via 🐍 v3.10.1 took 4s ❯ pdm add numpy==1.21.5 -v
Adding packages to default dependencies: numpy
======== Start resolving requirements ========
  numpy==1.21.5
  python
  Adding requirement numpy==1.21.5
	Found matching candidates:
	 <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>
  Adding requirement python
======== Starting round 0 ========
Resolving: new pin python 
======== Ending round 0 ========
======== Starting round 1 ========
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
  Adding requirement python>=3.7,<3.11(from numpy 1.21.5)
Conflicts detected: 
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
  python (from project)
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
Candidate rejected: python None
🔒 Lock failed
Unable to find a resolution because the following dependencies don't work on all Python versions defined by the project's `requires-python`: 
  python>=3.7,<3.11 (from <Candidate numpy 1.21.5 from https://pypi.org/simple/numpy/>)
To fix this, you can change the value of `requires-python` in pyproject.toml.
Traceback (most recent call last):
  File "/home/gregdan3/.local/bin/pdm", line 8, in <module>
    sys.exit(main())
  File "/home/gregdan3/.local/pipx/venvs/pdm/lib/python3.10/site-packages/pdm/core.py", line 219, in main
    return Core().main(args)
  File "/home/gregdan3/.local/pipx/venvs/pdm/lib/python3.10/site-packages/pdm/core.py", line 161, in main
    raise cast(Exception, err).with_traceback(traceback)
  File "/home/gregdan3/.local/pipx/venvs/pdm/lib/python3.10/site-packages/pdm/core.py", line 156, in main
    f(options.project, options)
  File "/home/gregdan3/.local/pipx/venvs/pdm/lib/python3.10/site-packages/pdm/cli/commands/add.py", line 53, in handle
    actions.do_add(
  File "/home/gregdan3/.local/pipx/venvs/pdm/lib/python3.10/site-packages/pdm/cli/actions.py", line 235, in do_add
    resolved = do_lock(project, strategy, tracked_names, reqs, dry_run=dry_run)
  File "/home/gregdan3/.local/pipx/venvs/pdm/lib/python3.10/site-packages/pdm/cli/actions.py", line 95, in do_lock
    raise ResolutionImpossible("Unable to find a resolution") from None
resolvelib.resolvers.ResolutionImpossible: Unable to find a resolution
```
The same goes for all the following `requires-python`: *, >=3.7, >=3.8, >=3.9, >=3.10.

As an aside, attempting to install numpy without a version pin fetches numpy==1.21.1 for requires-python>=3.10
```
~/temp2 via 🐍 v3.10.1 took 2s ❯ pdm add numpy -v
Adding packages to default dependencies: numpy
======== Start resolving requirements ========
  numpy
  python>=3.10
  Adding requirement numpy
	Found matching candidates:
	 <Candidate numpy 1.21.1 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.21.0 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.20.3 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.20.2 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.20.1 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.20.0 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.19.5 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.19.4 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.19.3 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.19.2 from https://pypi.org/simple/numpy/>
	 ... [1415 more candidate(s)]
  Adding requirement python>=3.10
======== Starting round 0 ========
Resolving: new pin python >=3.10
======== Ending round 0 ========
======== Starting round 1 ========
Resolving: new pin numpy 1.21.1
======== Ending round 1 ========
======== Starting round 2 ========
======== Resolution Result ========
Stable pins:
  python None
   numpy 1.21.1
	Found matching candidates:
	 <Candidate numpy 1.21.1 from https://pypi.org/simple/numpy/>
🔒 Lock successful
Changes are written to pdm.lock.
Changes are written to pyproject.toml.
Synchronizing working set with lock file: 1 to add, 0 to update, 0 to remove

Using cached wheel link: file:///home/gregdan3/.cache/pdm/wheels/cb/e9/ed/e7414af0de3ab84a43ebb26ea501fc425a1cecb1c2796e1eb6/numpy-1.21.1-cp310-cp310-linux_x86_64.whl
Install numpy 1.21.1 successful

🎉 All complete!
```

And using requires-python=* has the same result, but goes on to manually build numpy even though matching wheels should be available:
```
~/temp2 via 🐍 v3.10.1 ❯ pdm add numpy -v
Adding packages to default dependencies: numpy
======== Start resolving requirements ========
  numpy
  python
  Adding requirement numpy
	Found matching candidates:
	 <Candidate numpy 1.12.1 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.12.0 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.11.3 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.11.2 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.11.1 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.11.0 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.10.4 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.10.3 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.10.2 from https://pypi.org/simple/numpy/>
	 <Candidate numpy 1.10.1 from https://pypi.org/simple/numpy/>
	 ... [305 more candidate(s)]
  Adding requirement python
======== Starting round 0 ========
Resolving: new pin python 
======== Ending round 0 ========
======== Starting round 1 ========
Resolving: new pin numpy 1.12.1
======== Ending round 1 ========
======== Starting round 2 ========
======== Resolution Result ========
Stable pins:
  python None
   numpy 1.12.1
	Found matching candidates:
	 <Candidate numpy 1.12.1 from https://pypi.org/simple/numpy/>
🔒 Lock successful
Changes are written to pdm.lock.
Changes are written to pyproject.toml.
Synchronizing working set with lock file: 1 to add, 0 to update, 0 to remove

Preparing isolated env for PEP 517 build...
Collecting setuptools>=40.8.0
  Using cached setuptools-60.0.4-py3-none-any.whl (952 kB)
Collecting wheel
  Using cached wheel-0.37.1-py2.py3-none-any.whl (35 kB)
Installing collected packages: wheel, setuptools
  WARNING: The script wheel is installed in '/tmp/pdm-build-env-jibuu6zw-shared/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed setuptools-60.0.4 wheel-0.37.1
Running from numpy source directory.
running egg_info
creating numpy.egg-info
writing manifest file 'numpy.egg-info/SOURCES.txt'
warning: no previously-included files matching '*.pyc' found anywhere in distribution
warning: no previously-included files matching '*.pyo' found anywhere in distribution
warning: no previously-included files matching '*.pyd' found anywhere in distribution
writing manifest file 'numpy.egg-info/SOURCES.txt'
Running from numpy source directory.
/tmp/pdm-build-s6ouvbaf/numpy/distutils/misc_util.py:453: SyntaxWarning: "is" with a literal. Did you mean "=="?
  return is_string(s) and ('*' in s or '?' is s)
/tmp/pdm-build-s6ouvbaf/numpy/distutils/system_info.py:572: UserWarning: 
    Atlas (http://math-atlas.sourceforge.net/) libraries not found.
    Directories to search for the libraries can be specified in the
    numpy/distutils/site.cfg file (section [atlas]) or by setting
    the ATLAS environment variable.
  self.calc_info()
creating /tmp/tmp0x2m8n_4/tmp
creating /tmp/tmp0x2m8n_4/tmp/tmp0x2m8n_4
/bin/sh: line 1: svnversion: command not found
/bin/sh: line 1: svnversion: command not found
/tmp/pdm-build-env-jibuu6zw-shared/lib/python3.10/site-packages/setuptools/_distutils/dist.py:275: UserWarning: Unknown distribution option: 'define_macros'
  warnings.warn(msg)
non-existing path in 'numpy/distutils': 'site.cfg'
running bdist_wheel
running build
running config_cc
running config_fc
running build_src
```

## Expected behavior
numpy==1.21.5 should resolve and install successfully for at least `python-requires>=3.7` and `*`, as numpy on pypi distributes wheels for this major version and up.

## Environment Information
```bash
~/temp2 via 🐍 v3.10.1 ❯ pdm info && pdm info --env
PDM version:        1.12.0
Python Interpreter: /usr/bin/python (3.10)
Project Root:       /home/gregdan3/temp2
Project Packages:   /home/gregdan3/temp2/__pypackages__/3.10
{
  "implementation_name": "cpython",
  "implementation_version": "3.10.1",
  "os_name": "posix",
  "platform_machine": "x86_64",
  "platform_release": "5.15.7-arch1-1",
  "platform_system": "Linux",
  "platform_version": "#1 SMP PREEMPT Wed, 08 Dec 2021 14:33:16 +0000",
  "python_full_version": "3.10.1",
  "platform_python_implementation": "CPython",
  "python_version": "3.10",
  "sys_platform": "linux"
}
```

**Comment by gregdan3** [comment]
Yep; changing `requires-python` to `>=3.8,<3.11` worked as intended.

Why is it that resolution does not work when a matching version should be available? There is overlap between `>=3.7,<3.11` and `>=3.8`, so locking should be possible in theory.

**Comment by pawamoy** [comment]
Let say PDM lets you resolve. You publish your package. Someone tries to install it on Python 3.11. The Numpy dependency is fetched (if it is even feteched) and oops- it's not compatible with Python 3.11! pip/other refuses to install it.
